### ISSUE TITLE
StreamWrapper can not register protocol other than 's3'

### ISSUE THREAD

**Issue by dstevenson** [aws/aws-sdk-php#639]
I am using version 3.0.4 of the SDK

The `StreamWrapper::register` function accepts a `$protocol` parameter, however, if you pass in anything other than `s3` then `StreamWrapper::getClient()` will throw exception.

Issue is in `getOptions`

```
    private function getOptions($removeContextData = false)
    {
        // Context is not set when doing things like stat
        if ($this->context === null) {
            $options = [];
        } else {
            $options = stream_context_get_options($this->context);
            $options = isset($options['s3']) ? $options['s3'] : [];
        }

        $default = stream_context_get_options(stream_context_get_default());
        $default = isset($default['s3']) ? $default['s3'] : [];
        $result = $this->params + $options + $default;

        if ($removeContextData) {
            unset($result['client'], $result['seekable'], $result['cache']);
        }

        return $result;
    }
```

The initial `register` will set the `stream_context_get_default` with the correct key (the passed protocol), but `getOptions` is hardcoded to always check for `s3`. 

I would include failing unit tests but these are private functions

**Comment by jeremeamia** [comment]
Thanks for reporting this. We'll look into it.

**Comment by dstevenson** [comment]
@jeremeamia @mtdowling I am testing out the changes released for this bugfix in `3.0.7` and the protocol is not parsed when the streamwrapper executes a `rename` so it defaults to `s3` which will not exists in the default stream context.

Since we're working with 2 paths here I could see how that can be a little tricky since in theory the paths could have different protocols. Maybe it would be best to enforce that the `$path_from` and `$path_to` have the same protocol?
